Failed task should throw an exception

# Bug
## Summary
A failed task should fail properly

## Project Details
Version: `0.1.2`

## Description
The current sbt-docusaur finishes a task successfully even though it's actually failed. As a consequence, the failed task is not reported correctly to CI tools.

e.g.)
```sbt
sbt:sbt-docusaur> docusaurBuild
[error] Error when run npm for Docusaurus build at /path/to/sbt-docusaur/website
[error] Error when running the following command
[error]   command: npm run build
[error]   Error Code: ErrorCode(1)
[error]   Error:
[error]   - Error: Error while parsing markdown file /path/to/sbt-docusaur/website/docs/config.md
[error]   - Error while parsing markdown front matter.
[error]   - This can happen if you use special characteres like : in frontmatter values (try using "" around that value)
[error]   - unexpected end of the stream within a single quoted scalar at line 4, column 1:
[error]   -
[error]   -     ^
[error]   -     at Object.parseMarkdownFile (/path/to/sbt-docusaur/website/node_modules/@docusaurus/utils/lib/index.js:235:15)
...
[error]   - npm ERR! code ELIFECYCLE
[error]   - npm ERR! errno 1
[error]   - npm ERR! website@0.0.0 build: `docusaurus build`
[error]   - npm ERR! Exit status 1
[error]   - npm ERR!
[error]   - npm ERR! Failed at the website@0.0.0 build script.
[error]   - npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[error]   -
[error]   - npm ERR! A complete log of this run can be found in:
[error]   - npm ERR!     /path/to/.npm/_logs/2020-07-22T10_58_09_225Z-debug.log
[error]
[error]
[success] Total time: 2 s, completed 22 Jul. 2020, 8:58:09 pm
```
As shown above, it has `[success] Total time: 2 s, completed 22 Jul. 2020, 8:58:09 pm` at the end.

## Cause
The reason for the failed task being successfully done is that the failed task does not throw any exception.  Although it logs the error messages, it is not enough to tell the CI that it's failed since it returns no error code.

## Solution
A failed task should throw the `MessageOnlyException`.

--- a/src/main/scala/docusaur/Docusaur.scala
+++ b/src/main/scala/docusaur/Docusaur.scala
@@ -4,10 +4,12 @@ import java.io.File
 import java.nio.charset.Charset
 
 import cats._
+import cats.data.EitherT
 import cats.implicits._
 import docusaur.npm.Npm.NpmPath
 import docusaur.npm.{Npm, NpmCmd, NpmError}
 import effectie.Effectful._
+import effectie.cats.EitherTSupport._
 import effectie.cats.{CanCatch, EffectConstructor}
 import filef.{FileError2, FileF2}
 import loggerf.cats.Log.LeveledMessage._
@@ -21,36 +23,24 @@ import sbt.{IO => SbtIo}
  */
 object Docusaur {
 
-  def deleteFilesIn[F[_]: EffectConstructor: CanCatch: LogF: Monad](
+  def deleteFilesIn[F[_]: EffectConstructor: CanCatch: Monad](
     what: String,
     file: File
-  ): F[Unit] =
-    log(FileF2.deleteAllIn[F](file))(
-      err => error(
-        s"""Failed in deleting all files in ${file.getCanonicalPath} for $what
-           |  Error: ${FileError2.render(err)}
-           |""".stripMargin
-      ),
-      files => info(
-        s"""The following files are removed from ${file.getCanonicalPath}
-           |${files.mkString("    ", "\n    ", "\n")}
-           |""".stripMargin
-      )
-    ) *> effectOfUnit
+  ): F[Either[FileError2, List[String]]] =
+    FileF2.deleteAllIn[F](file)
 
   def install[F[_]: EffectConstructor: CanCatch: LogF: Monad](
-    what: String,
     npmPath: Option[NpmPath],
     docusaurusDir: File
-  ): F[Unit] =
-    runNpm("Docusaurus install", npmPath, docusaurusDir, NpmCmd.install)
+  ): F[Either[NpmError, Unit]] =
+    runAndLogNpm("Docusaurus install", npmPath, docusaurusDir, NpmCmd.install)
 
   def runBuild[F[_]: EffectConstructor: CanCatch: LogF: Monad](
     what: String,
     npmPath: Option[NpmPath],
     docusaurusDir: File
-  ): F[Unit] =
-    runNpm(
+  ): F[Either[NpmError, Unit]] =
+    runAndLogNpm(
       what,
       npmPath,
       docusaurusDir,
@@ -58,27 +48,22 @@ object Docusaur {
     )
 
 
-  def runNpm[F[_]: EffectConstructor: CanCatch: LogF: Monad](
+  def runAndLogNpm[F[_]: EffectConstructor: CanCatch: LogF: Monad](
     what: String,
     npmPath: Option[NpmPath],
     path: File,
     npmCmd: NpmCmd
-  ): F[Unit] =
-    log(Npm.run[F](npmPath, path.some, npmCmd))(
-      err => error(
-        s"""Error when run npm for $what at ${path.getCanonicalPath}
-           |${NpmError.render(err)}
+  ): F[Either[NpmError, Unit]] = (for {
+    result <-  EitherT(
+        Npm.run[F](npmPath, path.some, npmCmd)
+      )
+    _ <- eitherTRight[F, NpmError](LogF[F].logger0.info(
+        s"""Successfully run npm for $what
+           |  - command: npm run ${NpmCmd.values(npmCmd).mkString(" ")}
+           |${result.mkString("  ", "\n  ", "\n")}
            |""".stripMargin
-      ),
-      result =>
-        info(
-          s"""Successfully run npm for $what
-             |  - command: npm run ${NpmCmd.values(npmCmd).mkString(" ")}
-             |${result.mkString("  ", "\n  ", "\n")}
-             |""".stripMargin
-        )
-    ) *> effectOfUnit
-
+      ))
+  } yield ()).value
 
   @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
   def logAndWriteFile[F[_]: EffectConstructor: LogF: Monad](

--- a/src/main/scala/docusaur/DocusaurPlugin.scala
+++ b/src/main/scala/docusaur/DocusaurPlugin.scala
@@ -2,8 +2,11 @@ package docusaur
 
 import java.util.concurrent.{ConcurrentHashMap, ConcurrentMap}
 
+import cats.data.EitherT
 import cats.effect._
-import docusaur.npm.{Npm, NpmCmd}
+import docusaur.npm.{Npm, NpmCmd, NpmError}
+import effectie.cats.EitherTSupport._
+import filef.FileError2
 import githubpages.GitHubPagesPlugin
 import githubpages.GitHubPagesPlugin.{autoImport => ghpg}
 import loggerf.sbt.SbtLogger
@@ -36,23 +39,37 @@ object DocusaurPlugin extends AutoPlugin {
     }
   }
 
+  def toFileRemovalMessage(file: File, files: List[String]): String =
+    s"""The following files are removed from ${file.getCanonicalPath}
+       |${files.mkString("    ", "\n    ", "\n")}
+       |""".stripMargin
+
+  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+  def returnOrThrowMessageOnlyException[A, B](aOrB: Either[A, B])(aToString: A => String): B =
+    aOrB.fold(a => throw new MessageOnlyException(aToString(a)), identity)
+
 
   override lazy val projectSettings: Seq[Def.Setting[_]] = Seq(
 
     docusaurNpmPath := none[File],
 
     docusaurCleanNodeModules := Def.taskDyn {
-      @SuppressWarnings(Array("org.wartremover.warts.ExplicitImplicitTypes"))
-      implicit val logF = loggerFLogger(streams.value.log)
+      val log = streams.value.log
       val nodeModulesPath = docusaurDir.value / "node_modules"
         if (nodeModulesPath.exists()) {
           Def.task(
-            Docusaur.deleteFilesIn[IO]("'clean node_modules'", nodeModulesPath)
-              .unsafeRunSync()
+            returnOrThrowMessageOnlyException(
+              (for {
+                files <- EitherT(Docusaur.deleteFilesIn[IO]("'clean node_modules'", nodeModulesPath))
+                _ <- eitherTRight[IO, FileError2](log.info(toFileRemovalMessage(nodeModulesPath, files)))
+             } yield ())
+                .value
+                .unsafeRunSync()
+            )(FileError2.render)
           )
         } else {
           Def.task {
-            logF.info(s"node_modules does not exist at ${nodeModulesPath.getCanonicalPath}")
+            log.info(s"node_modules does not exist at ${nodeModulesPath.getCanonicalPath}")
             ()
           }
         }
@@ -64,8 +81,10 @@ object DocusaurPlugin extends AutoPlugin {
       val docusaurusDir = docusaurDir.value
       val npmPath = docusaurNpmPath.value.map(Npm.NpmPath)
       Def.task(
-        Docusaur.runNpm[IO]("Docusaurus install", npmPath, docusaurusDir, NpmCmd.install)
-          .unsafeRunSync()
+        returnOrThrowMessageOnlyException(
+          Docusaur.install[IO](npmPath, docusaurusDir)
+            .unsafeRunSync()
+        )(NpmError.render)
       )
     }.value,
 
@@ -75,8 +94,14 @@ object DocusaurPlugin extends AutoPlugin {
       val buildPath = docusaurBuildDir.value
       if (buildPath.exists()) {
         Def.task(
-          Docusaur.deleteFilesIn[IO]("'clean the Docusaurus build dir'", buildPath)
-            .unsafeRunSync()
+          returnOrThrowMessageOnlyException(
+            (for {
+              files <- EitherT(Docusaur.deleteFilesIn[IO] ("'clean the Docusaurus build dir'", buildPath))
+              _ <- eitherTRight[IO, FileError2](streams.value.log.info(toFileRemovalMessage(buildPath, files)))
+            } yield ())
+              .value
+              .unsafeRunSync()
+          )(FileError2.render)
         )
       } else {
         Def.task {
@@ -92,13 +117,15 @@ object DocusaurPlugin extends AutoPlugin {
       val docusaurusDir = docusaurDir.value
       val npmPath = docusaurNpmPath.value.map(Npm.NpmPath)
       Def.task(
-        Docusaur.runNpm[IO](
-          "Docusaurus build",
-          npmPath,
-          docusaurusDir,
-          NpmCmd.run(NpmCmd.Run.Param.build)
-        )
-        .unsafeRunSync()
+        returnOrThrowMessageOnlyException(
+          Docusaur.runAndLogNpm[IO](
+            "Docusaurus build",
+            npmPath,
+            docusaurusDir,
+            NpmCmd.run(NpmCmd.Run.Param.build)
+          )
+          .unsafeRunSync()
+        )(NpmError.render)
       )
     }.value,
 

